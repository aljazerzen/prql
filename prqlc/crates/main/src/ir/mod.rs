//! Intermediate Representations of Abstract Syntax Tree
//!
pub mod generic;
pub mod pl;
pub mod rq;
pub mod sm;

pub use prqlc_ast::Span;
