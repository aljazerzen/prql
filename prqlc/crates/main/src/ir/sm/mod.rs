//! Static minimal IR

mod expr;

pub use self::expr::*;
