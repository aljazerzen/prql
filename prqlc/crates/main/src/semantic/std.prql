# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Types

## Type primitives
type int
type float
type bool
type text
type date
type time
type timestamp
type `func`
type any

# Operators

let or = func left<bool> right<bool> -> <bool> internal std.or
let mul = func left right -> <int || float> internal std.mul
let div_i = func left right -> <int || float> internal std.div_i
let div_f = func left right -> <int || float> internal std.div_f
let mod = func left right -> <int || float> internal std.mod
let add = func left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.add
let sub = func left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.sub
let eq = func left right -> <bool> internal std.eq
let ne = func left right -> <bool> internal std.ne
let gt = func left right -> <bool> internal std.gt
let lt = func left right -> <bool> internal std.lt
let gte = func left right -> <bool> internal std.gte
let lte = func left right -> <bool> internal std.lte
let and = func left<bool> right<bool> -> <bool> internal std.and
let coalesce = func left right -> internal std.coalesce
let regex_search = func text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

## Generic array
type array = [any]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null

## Tuple of possibly many types
type tuple = {any..}

## Range
type range = {start = scalar, end = scalar}

## Relation (an array of tuples)
type relation = [tuple]

## Columnar relation (a tuple of arrays)
type tuple_of_arrays = {array..}

## Transform
type transform = (func relation -> relation)

# Functions

## Relational transforms
let from = func
  source <relation>
  -> <relation> source

@(implicit_closure on:0 param_names:[this])
let select = func
  columns <tuple>
  tbl <relation>
  -> <relation> internal select

@(implicit_closure on:0 param_names:[this])
let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

@(implicit_closure on:0 param_names:[this])
let derive = func
  columns <tuple>
  tbl <relation>
  -> <relation> internal derive

@(implicit_closure on:0 param_names:[this])
let aggregate = func
  columns <tuple>
  tbl <relation>
  -> <relation> internal aggregate

@(implicit_closure on:0 param_names:[this])
let sort = func
  by <tuple>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <scalar>
  tbl <relation>
  -> <relation> internal take

@(implicit_closure on:1 param_names:[this, that])
let join = func
  right <relation>
  condition <bool>
  `noresolve.side`:inner
  left <relation>
  -> <relation> internal join

@(implicit_closure on:0 param_names:[this])
let group = func
  by<scalar || tuple>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

@(implicit_closure on:0 param_names:[this])
let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  expr <tuple>
  tbl <relation>
  -> <relation> internal window

let append = bottom<relation> top<relation> -> <relation> internal append
# let intersect = bottom<relation> top<relation> -> <relation> (
#   top
#   join bottom (tuple_every (tuple_map _eq (tuple_zip this that)))
#   select this.top
# )
# let remove = bottom<relation> top<relation> -> <relation> (
#   top
#   join side:left (b = bottom) (tuple_every (tuple_map _eq (tuple_zip this that)))
#   filter (tuple_every (tuple_map _is_null b.*))
#   select this.top
# )
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let all = column <array> -> <bool> internal std.all

let some = column <array> -> <bool> internal std.some

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal std.count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@deprecated
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal std.row_number

## Misc functions
let round = n_digits column -> <scalar> internal std.round
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null
let tuple_exclude = func tuple except -> internal tuple_exclude

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## String functions
let lower = column -> <text> internal std.lower
let upper = column -> <text> internal std.upper

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv
