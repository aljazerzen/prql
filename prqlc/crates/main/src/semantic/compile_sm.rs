use anyhow::Result;
use prqlc_ast::error::Error;
use prqlc_ast::expr::Ident;

use crate::ir::{pl, sm};
use crate::semantic::Context;
use crate::WithErrorInfo;

pub fn compile_to_sm(context: Context, main_path: &[String]) -> Result<(sm::RootExpr, Context)> {
    // find main
    log::debug!("lookup for main pipeline in {main_path:?}");
    let (_, main_ident) = context.find_main_rel(main_path).map_err(|hint| {
        Error::new_simple("Missing main pipeline")
            .with_code("E0001")
            .with_hints(hint)
    })?;

    // find & validate query def
    // let def = context.find_query_def(&main_ident);
    // let def = def.cloned().unwrap_or_default();
    // validate query def

    let mut compiler = SmCompiler {
        context,
        exprs: Vec::new(),
    };

    let _root_eid = compiler.compile_decl_as_expr(main_ident)?;

    let root_expr = sm::RootExpr {
        exprs: compiler.exprs,
    };

    Ok((root_expr, compiler.context))
}

struct SmCompiler {
    context: Context,
    exprs: Vec<sm::Expr>,
}

impl SmCompiler {
    fn compile_decl_as_expr(&mut self, ident: Ident) -> Result<sm::EId> {
        let decl = self.context.root_mod.get(&ident).unwrap();

        let expr = (decl.kind.as_expr())
            .ok_or_else(|| Error::new_simple(format!("expected `{ident}` to be an expression")))?;

        self.compile_expr(*expr.clone())
    }

    fn compile_expr(&mut self, expr: pl::Expr) -> Result<sm::EId> {
        let kind = match expr.kind {
            pl::ExprKind::Ident(ident) => todo!("ident: {ident}"),
            pl::ExprKind::Literal(lit) => sm::ExprKind::Literal(lit),

            pl::ExprKind::Tuple(fields) => sm::ExprKind::Tuple(self.compile_exprs(fields)?),
            pl::ExprKind::Array(items) => sm::ExprKind::Array(self.compile_exprs(items)?),

            pl::ExprKind::RqOperator { name, args } => {
                let args = self.compile_exprs(args)?;
                sm::ExprKind::Operator { name, args }
            }

            pl::ExprKind::Param(param_id) => sm::ExprKind::Param(param_id),

            pl::ExprKind::Func(_) => todo!(),

            pl::ExprKind::TupleFields(_) => todo!(),
            pl::ExprKind::TupleExclude { .. } => todo!(),

            pl::ExprKind::SString(_) => todo!(),
            pl::ExprKind::Case(_) => todo!(),

            pl::ExprKind::TransformCall(_) => todo!("this pl::ExprKind should be removed"),

            pl::ExprKind::FuncCall(_)
            | pl::ExprKind::Type(_)
            | pl::ExprKind::Internal(_)
            | pl::ExprKind::FString(_) => {
                unreachable!()
            }
        };

        let expr = sm::Expr {
            kind,
            ty: expr.ty.expect("expression type to be resolved"),
            span: expr.span,
        };

        let eid = sm::EId::from(self.exprs.len() as u32);
        self.exprs.push(expr);

        Ok(eid)
    }

    fn compile_exprs(&mut self, exprs: Vec<pl::Expr>) -> Result<Vec<sm::EId>> {
        exprs.into_iter().map(|e| self.compile_expr(e)).collect()
    }
}
