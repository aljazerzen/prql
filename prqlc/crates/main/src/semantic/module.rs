use std::collections::HashMap;

use anyhow::Result;
use prqlc_ast::stmt::QueryDef;

use crate::ir::decl::*;
use crate::ir::pl::{Annotation, Expr, Ident, Ty, TyKind, TyTuple};
use crate::Error;

use super::{
    NS_DEFAULT_DB, NS_INFER, NS_INFER_MODULE, NS_MAIN, NS_PARAM, NS_QUERY_DEF, NS_SELF, NS_STD,
    NS_THAT, NS_THIS,
};

impl Module {
    pub fn singleton<S: ToString>(name: S, entry: Decl) -> Module {
        Module {
            names: HashMap::from([(name.to_string(), entry)]),
            ..Default::default()
        }
    }

    pub fn new_root() -> Module {
        // Each module starts with a default namespace that contains a wildcard
        // and the standard library.
        Module {
            names: HashMap::from([
                (
                    NS_DEFAULT_DB.to_string(),
                    Decl::from(DeclKind::Module(Self::new_database())),
                ),
                (NS_STD.to_string(), Decl::from(DeclKind::default())),
            ]),
            shadowed: None,
            redirects: [
                Ident::from_name(NS_THIS),
                Ident::from_name(NS_THAT),
                Ident::from_name(NS_PARAM),
                Ident::from_name(NS_STD),
            ]
            .into(),
        }
    }

    pub fn new_database() -> Module {
        let mut col_ty = Ty::new(TyKind::Any);
        col_ty.infer = true;

        let table_ty = Ty::relation(TyTuple {
            fields: Vec::new(),
            has_other: true,
        });

        let names = HashMap::from([
            (
                NS_INFER.to_string(),
                Decl::from(DeclKind::Infer(Box::new(DeclKind::Expr(Box::new(Expr {
                    ty: Some(table_ty),
                    ..Expr::new(Ident::from_name(NS_INFER))
                }))))),
            ),
            (
                NS_INFER_MODULE.to_string(),
                Decl::from(DeclKind::Infer(Box::new(DeclKind::Module(Module {
                    names: HashMap::new(),
                    redirects: [].into(),
                    shadowed: None,
                })))),
            ),
        ]);
        Module {
            names,
            shadowed: None,
            redirects: [].into(),
        }
    }

    pub fn insert(&mut self, fq_ident: Ident, decl: Decl) -> Result<Option<Decl>, Error> {
        if fq_ident.path.is_empty() {
            Ok(self.names.insert(fq_ident.name, decl))
        } else {
            let (top_level, remaining) = fq_ident.pop_front();
            let entry = self.names.entry(top_level).or_default();

            if let DeclKind::Module(inner) = &mut entry.kind {
                inner.insert(remaining.unwrap(), decl)
            } else {
                Err(Error::new_simple(
                    "path does not resolve to a module or a table",
                ))
            }
        }
    }

    pub fn get_mut(&mut self, ident: &Ident) -> Option<&mut Decl> {
        let mut ns = self;

        for part in &ident.path {
            let entry = ns.names.get_mut(part);

            match entry {
                Some(Decl {
                    kind: DeclKind::Module(inner),
                    ..
                }) => {
                    ns = inner;
                }
                _ => return None,
            }
        }

        ns.names.get_mut(&ident.name)
    }

    /// Get namespace entry using a fully qualified ident.
    pub fn get(&self, fq_ident: &Ident) -> Option<&Decl> {
        let mut ns = self;

        for (index, part) in fq_ident.path.iter().enumerate() {
            let decl = ns.names.get(part);
            if let Some(decl) = decl {
                match &decl.kind {
                    DeclKind::Module(inner) => {
                        ns = inner;
                    }
                    DeclKind::LayeredModules(stack) => {
                        let next = fq_ident.path.get(index + 1).unwrap_or(&fq_ident.name);
                        let mut found = false;
                        for n in stack.iter().rev() {
                            if n.names.contains_key(next) {
                                ns = n;
                                found = true;
                                break;
                            }
                        }
                        if !found {
                            return None;
                        }
                    }
                    _ => return None,
                }
            } else {
                return None;
            }
        }

        ns.names.get(&fq_ident.name)
    }

    pub(super) fn insert_relation(&mut self, expr: &Expr, namespace: &str) {
        let ty = expr.ty.as_ref().unwrap();
        let tuple = ty.kind.as_array().unwrap();

        self.insert_ty(namespace.to_string(), tuple, 0);
    }

    /// Creates a name resolution declaration that allows lookups into the given type.
    fn insert_ty(&mut self, name: String, ty: &Ty, order: usize) {
        log::debug!("inserting `{name}`: {ty}");

        let decl_kind = match &ty.kind {
            // for tuples, create a submodule
            TyKind::Tuple(tuple) => {
                let mut sub_mod = Module::default();

                if let Some(instance_of) = &ty.instance_of {
                    let lineage = ty.lineage.unwrap();

                    let self_decl = Decl {
                        declared_at: None,
                        kind: DeclKind::InstanceOf {
                            table_fq: instance_of.clone(),
                            lineage,
                        },
                        ..Default::default()
                    };
                    sub_mod.names.insert(NS_SELF.to_string(), self_decl);
                }

                for (index, (name, ty)) in tuple.fields.iter().enumerate() {
                    if let Some(name) = name {
                        sub_mod.insert_ty(name.clone(), ty.as_ref().unwrap(), index + 1);
                    } else {
                        // unnamed tuple fields cannot be references,
                        // so there is no point of having them in the module
                    }
                }
                if tuple.has_other {
                    let field_ty = Ty {
                        lineage: ty.lineage,
                        ..Ty::new(TyKind::Any)
                    };

                    let decl_kind = DeclKind::Infer(Box::new(DeclKind::Column(field_ty)));

                    let mut decl = Decl::from(decl_kind);
                    decl.order = tuple.fields.len() + 1;
                    sub_mod.names.insert(NS_INFER.to_string(), decl);
                }

                self.redirects.insert(Ident::from_name(&name));
                DeclKind::Module(sub_mod)
            }

            // for anything else, create a plain column
            _ => DeclKind::Column(ty.clone()),
        };

        let mut decl = Decl::from(decl_kind);
        decl.order = order;
        self.names.insert(name, decl);
    }

    pub fn shadow(&mut self, ident: &str) {
        let shadowed = self.names.remove(ident).map(Box::new);
        let entry = DeclKind::Module(Module {
            shadowed,
            ..Default::default()
        });
        self.names.insert(ident.to_string(), entry.into());
    }

    pub fn unshadow(&mut self, ident: &str) {
        if let Some(entry) = self.names.remove(ident) {
            let ns = entry.kind.into_module().unwrap();

            if let Some(shadowed) = ns.shadowed {
                self.names.insert(ident.to_string(), *shadowed);
            }
        }
    }

    pub fn stack_push(&mut self, ident: &str, namespace: Module) {
        let entry = self
            .names
            .entry(ident.to_string())
            .or_insert_with(|| DeclKind::LayeredModules(Vec::new()).into());
        let stack = entry.kind.as_layered_modules_mut().unwrap();

        stack.push(namespace);
    }

    pub fn stack_pop(&mut self, ident: &str) -> Option<Module> {
        (self.names.get_mut(ident))
            .and_then(|e| e.kind.as_layered_modules_mut())
            .and_then(|stack| stack.pop())
    }

    pub(crate) fn into_exprs(self) -> HashMap<String, Expr> {
        self.names
            .into_iter()
            .map(|(k, v)| (k, *v.kind.into_expr().unwrap()))
            .collect()
    }

    pub(crate) fn from_exprs(exprs: HashMap<String, Expr>) -> Module {
        Module {
            names: exprs
                .into_iter()
                .map(|(key, expr)| {
                    let decl = Decl {
                        kind: DeclKind::Expr(Box::new(expr)),
                        ..Default::default()
                    };
                    (key, decl)
                })
                .collect(),
            ..Default::default()
        }
    }

    pub fn as_decls(&self) -> Vec<(Ident, &Decl)> {
        let mut r = Vec::new();
        for (name, decl) in &self.names {
            match &decl.kind {
                DeclKind::Module(module) => r.extend(
                    module
                        .as_decls()
                        .into_iter()
                        .map(|(inner, decl)| (Ident::from_name(name) + inner, decl)),
                ),
                _ => r.push((Ident::from_name(name), decl)),
            }
        }
        r
    }

    /// Recursively finds all declarations that end in suffix.
    pub fn find_by_suffix(&self, suffix: &str) -> Vec<Ident> {
        let mut res = Vec::new();

        for (name, decl) in &self.names {
            if let DeclKind::Module(module) = &decl.kind {
                let nested = module.find_by_suffix(suffix);
                res.extend(nested.into_iter().map(|x| x.prepend(vec![name.clone()])));
                continue;
            }

            if name == suffix {
                res.push(Ident::from_name(name));
            }
        }

        res
    }
}

impl RootModule {
    pub(super) fn declare(
        &mut self,
        ident: Ident,
        decl: DeclKind,
        id: Option<usize>,
        annotations: Vec<Annotation>,
    ) -> Result<()> {
        let existing = self.module.get(&ident);
        if existing.is_some() {
            return Err(Error::new_simple(format!("duplicate declarations of {ident}")).into());
        }

        let decl = Decl {
            kind: decl,
            declared_at: id,
            order: 0,
            annotations,
        };
        self.module.insert(ident, decl).unwrap();
        Ok(())
    }

    /// Finds that main pipeline given a path to either main itself or its parent module.
    /// Returns main expr and fq ident of the decl.
    pub fn find_main_rel(&self, path: &[String]) -> Result<(&Expr, Ident), Option<String>> {
        let (decl, ident) = self.find_main(path)?;

        let decl = (decl.kind.as_expr()).ok_or(Some(format!("{ident} is not an expression")))?;

        Ok((decl, ident))
    }

    pub fn find_main(&self, path: &[String]) -> Result<(&Decl, Ident), Option<String>> {
        let mut tried_idents = Vec::new();

        // is path referencing the relational var directly?
        if !path.is_empty() {
            let ident = Ident::from_path(path.to_vec());
            let decl = self.module.get(&ident);

            if let Some(decl) = decl {
                return Ok((decl, ident));
            } else {
                tried_idents.push(ident.to_string());
            }
        }

        // is path referencing the parent module?
        {
            let mut path = path.to_vec();
            path.push(NS_MAIN.to_string());

            let ident = Ident::from_path(path);
            let decl = self.module.get(&ident);

            if let Some(decl) = decl {
                return Ok((decl, ident));
            } else {
                tried_idents.push(ident.to_string());
            }
        }

        Err(Some(format!(
            "Expected a declaration at {}",
            tried_idents.join(" or ")
        )))
    }

    pub fn find_query_def(&self, main: &Ident) -> Option<&QueryDef> {
        let ident = Ident {
            path: main.path.clone(),
            name: NS_QUERY_DEF.to_string(),
        };

        let decl = self.module.get(&ident)?;
        decl.kind.as_query_def()
    }

    /// Finds all main pipelines.
    pub fn find_mains(&self) -> Vec<Ident> {
        self.module.find_by_suffix(NS_MAIN)
    }

    pub(super) fn find_std_type(&self, name: &str) -> Ty {
        let ident = Ident::from_path(vec![NS_STD, name]);
        let decl = self.module.get(&ident).unwrap().kind.as_expr().unwrap();
        decl.kind.as_type().unwrap().clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ir::pl::{Expr, ExprKind, Literal};

    #[test]
    fn test_module_shadow_unshadow() {
        let mut module = Module::default();

        let ident = Ident::from_name("test_name");
        let expr: Expr = Expr::new(ExprKind::Literal(Literal::Integer(42)));
        let decl: Decl = DeclKind::Expr(Box::new(expr)).into();

        module.insert(ident.clone(), decl.clone()).unwrap();

        module.shadow("test_name");
        assert!(module.get(&ident) != Some(&decl));

        module.unshadow("test_name");
        assert_eq!(module.get(&ident).unwrap(), &decl);
    }
}
