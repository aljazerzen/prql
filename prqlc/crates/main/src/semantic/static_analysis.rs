//! Static analysis - compile time expression evaluation

use crate::ir::pl::{Expr, ExprKind, Literal};

pub fn static_analysis(mut expr: Expr) -> Expr {
    match expr.kind {
        ExprKind::RqOperator { name, mut args } => {
            match name.as_str() {
                "std.not" => {
                    if let ExprKind::Literal(Literal::Boolean(val)) = &args[0].kind {
                        return Expr::new(Literal::Boolean(!val));
                    }
                }
                "std.neg" => match &args[0].kind {
                    ExprKind::Literal(Literal::Integer(val)) => {
                        return Expr::new(Literal::Integer(-val))
                    }
                    ExprKind::Literal(Literal::Float(val)) => {
                        return Expr::new(Literal::Float(-val))
                    }
                    _ => (),
                },

                "std.eq" => {
                    if let (ExprKind::Literal(left), ExprKind::Literal(right)) =
                        (&args[0].kind, &args[1].kind)
                    {
                        // don't eval comparisons between different types of literals
                        if left.as_ref() == right.as_ref() {
                            return Expr::new(Literal::Boolean(left == right));
                        }
                    }
                }
                "std.ne" => {
                    if let (ExprKind::Literal(left), ExprKind::Literal(right)) =
                        (&args[0].kind, &args[1].kind)
                    {
                        // don't eval comparisons between different types of literals
                        if left.as_ref() == right.as_ref() {
                            return Expr::new(Literal::Boolean(left != right));
                        }
                    }
                }
                "std.and" => {
                    if let (
                        ExprKind::Literal(Literal::Boolean(left)),
                        ExprKind::Literal(Literal::Boolean(right)),
                    ) = (&args[0].kind, &args[1].kind)
                    {
                        return Expr::new(Literal::Boolean(*left && *right));
                    }
                }
                "std.or" => {
                    if let (
                        ExprKind::Literal(Literal::Boolean(left)),
                        ExprKind::Literal(Literal::Boolean(right)),
                    ) = (&args[0].kind, &args[1].kind)
                    {
                        return Expr::new(Literal::Boolean(*left || *right));
                    }
                }
                "std.coalesce" => {
                    if let ExprKind::Literal(Literal::Null) = &args[0].kind {
                        return args.remove(1);
                    }
                }

                _ => {}
            };
            expr.kind = ExprKind::RqOperator { name, args };
            expr
        }

        ExprKind::Case(items) => {
            let mut res = Vec::with_capacity(items.len());
            for item in items {
                if let ExprKind::Literal(Literal::Boolean(condition)) = item.condition.kind {
                    if condition {
                        res.push(item);
                        break;
                    } else {
                        // this case can be removed
                        continue;
                    }
                } else {
                    res.push(item);
                }
            }
            if res.is_empty() {
                return Expr::new(Literal::Null);
            }

            if res.len() == 1 {
                let is_true = matches!(
                    res[0].condition.kind,
                    ExprKind::Literal(Literal::Boolean(true))
                );
                if is_true {
                    return *res.remove(0).value;
                }
            }

            expr.kind = ExprKind::Case(res);
            expr
        }

        _ => expr,
    }
}
