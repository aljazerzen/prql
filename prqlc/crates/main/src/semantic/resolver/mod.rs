mod expr;
mod flatten;
mod functions;
mod name_resolution;
mod static_eval;
mod stmt;
mod transforms;
mod types;

use crate::ir::decl::RootModule;
use crate::semantic::{NS_PARAM, NS_STD};
use crate::utils::IdGenerator;

/// Can fold (walk) over AST and for each function call or variable find what they are referencing.
pub struct Resolver {
    pub root_mod: RootModule,

    pub current_module_path: Vec<String>,

    /// Sometimes ident closures must be resolved and sometimes not. See [test::test_func_call_resolve].
    pub(super) in_func_call_name: bool,

    disable_type_checking: bool,

    pub id: IdGenerator<usize>,

    pub options: ResolverOptions,
}

#[derive(Default, Clone)]
pub struct ResolverOptions {
    pub allow_module_decls: bool,
}

impl Resolver {
    pub fn new(root_mod: RootModule, options: ResolverOptions) -> Self {
        Resolver {
            root_mod,
            options,
            current_module_path: Vec::new(),
            in_func_call_name: false,
            disable_type_checking: false,
            id: IdGenerator::new(),
        }
    }
}

#[cfg(test)]
pub(super) mod test {
    use anyhow::Result;
    use insta::assert_yaml_snapshot;

    use crate::ir::pl::{Expr, PlFold, Ty};

    pub fn erase_ids(expr: Expr) -> Expr {
        IdEraser {}.fold_expr(expr).unwrap()
    }

    struct IdEraser {}

    impl PlFold for IdEraser {
        fn fold_expr(&mut self, mut expr: Expr) -> Result<Expr> {
            expr.kind = self.fold_expr_kind(expr.kind)?;
            expr.id = None;
            expr.target_id = None;
            Ok(expr)
        }
    }

    fn parse_and_resolve(query: &str) -> Result<Expr> {
        let ctx = crate::semantic::test::parse_and_resolve(query)?;
        let (main, _) = ctx.find_main_rel(&[]).unwrap();
        Ok(main.clone())
    }

    fn resolve_ty(query: &str) -> Result<Ty> {
        Ok(parse_and_resolve(query)?.ty.unwrap())
    }

    fn resolve_derive(query: &str) -> Result<Vec<Expr>> {
        let expr = parse_and_resolve(query)?;
        let derive = expr.kind.into_transform_call().unwrap();
        let exprs = derive
            .kind
            .into_derive()
            .unwrap_or_else(|e| panic!("Failed to convert `{e:?}`"))
            .kind
            .into_tuple()
            .unwrap_or_else(|e| panic!("Failed to convert `{e:?}`"));
        let exprs = IdEraser {}.fold_exprs(exprs).unwrap();
        Ok(exprs)
    }

    #[test]
    fn test_variables_1() {
        assert_yaml_snapshot!(resolve_derive(
            r#"
            from employees
            derive {
                gross_salary = salary + payroll_tax,
                gross_cost =   gross_salary + benefits_cost
            }
            "#
        )
        .unwrap());
    }

    #[test]
    fn test_non_existent_function() {
        parse_and_resolve(r#"from mytable | filter (myfunc col1)"#).unwrap_err();
    }

    #[test]
    fn test_functions_1() {
        assert_yaml_snapshot!(resolve_derive(
            r#"
            let subtract = a b -> a - b

            from employees
            derive {
                net_salary = subtract gross_salary tax
            }
            "#
        )
        .unwrap());
    }

    #[test]
    fn test_functions_nested() {
        assert_yaml_snapshot!(resolve_derive(
            r#"
            let lag_day = x -> s"lag_day_todo({x})"
            let ret = x dividend_return ->  x / (lag_day x) - 1 + dividend_return

            from a
            derive (ret b c)
            "#
        )
        .unwrap());
    }

    #[test]
    fn test_functions_pipeline() {
        assert_yaml_snapshot!(resolve_derive(
            r#"
            from a
            derive one = (foo | sum)
            "#
        )
        .unwrap());

        assert_yaml_snapshot!(resolve_derive(
            r#"
            let plus_one = x -> x + 1
            let plus = x y -> x + y

            from a
            derive {b = (sum foo | plus_one | plus 2)}
            "#
        )
        .unwrap());
    }
    #[test]
    fn test_named_args() {
        assert_yaml_snapshot!(resolve_derive(
            r#"
            let add_one = x to:1 -> x + to

            from foo_table
            derive {
                added = add_one bar to:3,
                added_default = add_one bar
            }
            "#
        )
        .unwrap());
    }

    #[test]
    fn test_frames_and_names() {
        assert_yaml_snapshot!(resolve_ty(
            r#"
            from orders
            select {customer_no, gross, tax, gross - tax}
            take 20
            "#
        )
        .unwrap());

        assert_yaml_snapshot!(resolve_ty(
            r#"
            from table_1
            join customers (==customer_no)
            "#
        )
        .unwrap());

        assert_yaml_snapshot!(resolve_ty(
            r#"
            from e = employees
            join salaries (==emp_no)
            group {e.emp_no, e.gender} (
                aggregate {
                    emp_salary = average salaries.salary
                }
            )
            "#
        )
        .unwrap());
    }

    #[test]
    fn test_tuple_field_name_inference() {
        // test that:
        // - type contains aliases,
        // - names are not duplicated, but overridden,
        // - when using plain ident without alias, the ident is used as alias (TODO)
        assert_yaml_snapshot!(resolve_ty(
            r#"
            [{a = 1, b = 2, a = 3}]
            "#
        )
        .unwrap(), @r###"
        ---
        kind:
          Array:
            kind:
              Tuple:
                - Single:
                    - ~
                    - kind:
                        Primitive: Int
                      lineage: 180
                - Single:
                    - b
                    - kind:
                        Primitive: Int
                      lineage: 181
                - Single:
                    - a
                    - kind:
                        Primitive: Int
                      lineage: 182
            lineage: 179
        lineage: 178
        "###);
    }

    #[test]
    fn test_tuple_field_exclude() {
        assert_yaml_snapshot!(resolve_ty(
            r#"
            from tracks
            select (std.tuple_exclude {track_id, milliseconds, title, price} {milliseconds, title})
            "#
        )
        .unwrap(), @r###"
        ---
        kind:
          Array:
            kind:
              Tuple:
                - Single:
                    - track_id
                    - kind: Any
                      lineage: 203
                      infer: true
                - Single:
                    - price
                    - kind: Any
                      lineage: 203
                      infer: true
            lineage: 210
        "###);

        assert_yaml_snapshot!(resolve_ty(
            r#"
            from tracks
            select !{title, composer}
            "#
        )
        .unwrap(), @r###"
        ---
        kind:
          Array:
            kind:
              Tuple:
                - All:
                    ty:
                      kind: Any
                      lineage: 203
                      instance_of:
                        - default_db
                        - tracks
                      infer: true
                    exclude:
                      - - title
                      - - composer
            lineage: 203
            instance_of:
              - default_db
              - tracks
        "###);
    }
}
