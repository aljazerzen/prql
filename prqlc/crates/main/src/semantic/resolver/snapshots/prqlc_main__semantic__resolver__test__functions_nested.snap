---
source: prqlc/crates/main/src/semantic/resolver/mod.rs
expression: "resolve_derive(r#\"\n            let lag_day = x -> s\"lag_day_todo({x})\"\n            let ret = x dividend_return ->  x / (lag_day x) - 1 + dividend_return\n\n            from a\n            derive (ret b c)\n            \"#).unwrap()"
---
- RqOperator:
    name: std.add
    args:
      - RqOperator:
          name: std.sub
          args:
            - RqOperator:
                name: std.div_f
                args:
                  - Ident:
                      - this
                      - a
                      - b
                    ty:
                      kind: Any
                      lineage: 205
                      infer: true
                  - SString:
                      - String: lag_day_todo(
                      - Expr:
                          expr:
                            Ident:
                              - this
                              - a
                              - b
                            ty:
                              kind: Any
                              lineage: 205
                              infer: true
                          format: ~
                      - String: )
                    ty:
                      kind:
                        Union:
                          - - int
                            - kind:
                                Primitive: Int
                              name: int
                          - - float
                            - kind:
                                Primitive: Float
                              name: float
                          - - bool
                            - kind:
                                Primitive: Bool
                              name: bool
                          - - text
                            - kind:
                                Primitive: Text
                              name: text
                          - - date
                            - kind:
                                Primitive: Date
                              name: date
                          - - time
                            - kind:
                                Primitive: Time
                              name: time
                          - - timestamp
                            - kind:
                                Primitive: Timestamp
                              name: timestamp
                          - - ~
                            - kind:
                                Singleton: "Null"
                      name: scalar
              ty:
                kind:
                  Union:
                    - - int
                      - kind:
                          Primitive: Int
                        name: int
                    - - float
                      - kind:
                          Primitive: Float
                        name: float
            - Literal:
                Integer: 1
              ty:
                kind:
                  Primitive: Int
                lineage: 426
        ty:
          kind:
            Union:
              - - int
                - kind:
                    Primitive: Int
                  name: int
              - - float
                - kind:
                    Primitive: Float
                  name: float
              - - timestamp
                - kind:
                    Primitive: Timestamp
                  name: timestamp
              - - date
                - kind:
                    Primitive: Date
                  name: date
      - Ident:
          - this
          - a
          - c
        ty:
          kind:
            Union:
              - - int
                - kind:
                    Primitive: Int
                  name: int
              - - float
                - kind:
                    Primitive: Float
                  name: float
              - - timestamp
                - kind:
                    Primitive: Timestamp
                  name: timestamp
              - - date
                - kind:
                    Primitive: Date
                  name: date
  ty:
    kind:
      Union:
        - - int
          - kind:
              Primitive: Int
            name: int
        - - float
          - kind:
              Primitive: Float
            name: float
        - - timestamp
          - kind:
              Primitive: Timestamp
            name: timestamp
        - - date
          - kind:
              Primitive: Date
            name: date

